tiledb/sm/serialization/query.cc

```diff
@@ -261,84 +261,122 @@ Status subarray_to_capnp(
-  RETURN_NOT_OK(subarray->set_coalesce_ranges(reader.getCoalesceRanges()));
+Subarray subarray_from_capnp(
+    const capnp::Subarray::Reader& reader,
+    const Array* array,
+    Layout layout,
+    stats::Stats* parent_stats,
+    shared_ptr<Logger> logger) {
+  bool coalesce_ranges = reader.getCoalesceRanges();
   auto ranges_reader = reader.getRanges();
+
   uint32_t dim_num = ranges_reader.size();
+  std::vector<RangeSetAndSuperset> range_subset;
+  range_subset.reserve(dim_num);
+  std::vector<bool> is_default(dim_num, false);
   for (uint32_t i = 0; i < dim_num; i++) {
     auto range_reader = ranges_reader[i];
-
-    auto data = range_reader.getBuffer();
-    auto data_ptr = data.asBytes();
-    if (range_reader.hasBufferSizes()) {
-      auto ranges = range_buffers_from_capnp(range_reader);
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, ranges));
-
-      // Set default indicator
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+    Datatype type = datatype_enum(range_reader.getType());
+    auto dim = array->array_schema_latest().dimension_ptr(i);
+
+    is_default[i] = range_reader.getHasDefaultRange();
+    if (is_default[i]) {
+      // If the range is implicitly initialized, the RangeSetAndSuperset
+      // constructor will initialize the ranges to the domain.
+      range_subset.emplace_back(
+          type, dim->domain(), is_default[i], coalesce_ranges);
     } else {
-      // Handle 1.7 style ranges where there is a single range with no sizes
-      Range range(data_ptr.begin(), data.size());
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, {range}));
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+      std::vector<Range> ranges;
+      if (range_reader.hasBufferSizes()) {
+        ranges = range_buffers_from_capnp(range_reader);
+        // Add custom ranges, clearing any implicit ranges previously set.
+        range_subset.emplace_back(
+            type, dim->domain(), std::move(ranges), coalesce_ranges);
+      } else {
+        // Handle 1.7 style ranges where there is a single range with no sizes
+        auto data = range_reader.getBuffer();
+        auto data_ptr = data.asBytes();
+        ranges.emplace_back(data_ptr.begin(), data.size());
+        range_subset.emplace_back(
+            type, dim->domain(), std::move(ranges), coalesce_ranges);
+      }
     }
   }
 
+  std::vector<optional<Subarray::LabelRangeSubset>> label_range_subset;
+  label_range_subset.reserve(dim_num);
+  uint32_t last_dim = 0;
   if (reader.hasLabelRanges()) {
-    subarray->add_default_label_ranges(dim_num);
     auto label_ranges_reader = reader.getLabelRanges();
     uint32_t label_num = label_ranges_reader.size();
     for (uint32_t i = 0; i < label_num; i++) {
       auto label_range_reader = label_ranges_reader[i];
-      auto dim_id = label_range_reader.getDimensionId();
+      auto dim_index = label_range_reader.getDimensionId();
```

Comment:
Same comment here. It would be better to have `label_subarray_ranges_from_capnp`.